[MRG+2] Adding Implementation of SAG - next episode

TomDLT opened the pull request:
I took over the great work of @dsullivan7 in #3814.

I removed the merges with master, squashed all the commits and rebased on master.

dsullivan7 committed:
Adding Implementation of SAG

TomDLT committed:
ENH add sag solver in LogisticRegression and Ridge

ENH add new parameter random_state in Ridge class
ENH add new parameter warm_start in LogisticRegression

dsullivan7 commented:
Awesome @TomDLT! There was talk of having this implemented as a solver for LogisticRegression and RidgeRegression, have you looked into that at all?

amueller commented:
travis is still unhappy ;) Thanks for picking this up!

TomDLT commented:
I rerun the classifier benchmark on two large datasets:
RCV1 and Alpha [(cf here)](http://leon.bottou.org/projects/sgd)
The plot shows the convergence with `log10(|loss - loss_optimal|)`

Result on Alpha (500.000 x 500, Dense):
![diffloss](https://cloud.githubusercontent.com/assets/11065596/7704660/0aaab458-fe40-11e4-98b8-047a79e328a3.png)
Result on RCV1 (804.414 x 47.152, Sparse):
![diffloss](https://cloud.githubusercontent.com/assets/11065596/7698883/14b50d8a-fe14-11e4-95ac-be6e57db78a3.png)

agramfort commented:
as discussed I vote for adding 'sag' solver to LogisticRegression and RidgeRegression that would call plain sag_logistic and sag_ridge functions.

amueller commented:
newtoncg is faster than liblinear? I'm surprised! Anyhow SAG seems to kick ass. I'd be +1 on adding a solver to the classifiers as this seems like a good default.

amueller commented:
(though i dream of the day where the default LogisticRegression is multinomial, not OvR ;)

TomDLT commented:
>newton-cg is faster than liblinear? I'm surprised!

Actually, newton-cg is not faster. 
In previous example, with `fit_intercept=True`, liblinear and newton-cg does not converge to the same minimum, since liblinear use a regularization on the intercept, whereas newton-cg and SAG don't.

I tried using the same regularization in SAG, and it converges to the same minimum as liblinear.
However, it makes more sense not to regularize the intercept.

Finally, with `fit_intercept=False`, we see that liblinear is a not slower than newton-cg.
![diffloss_no_intercept](https://cloud.githubusercontent.com/assets/11065596/7707388/cc0a9de8-fe4f-11e4-9f42-72d2513f541a.png)

amueller commented:
Thanks for the explanation.

TomDLT commented:
I implemented `sag_logistic` as a solver in `LogisticRegression`, and changed accordingly some of the tests.

Currently, in order to match `LogisticRegression` API, and compared to previous `SAGClassifier`, 
- `eta` is forced to `'auto'`
- we loose the `warm_start`
- we loose parallel processing for multiclass
- the behavior with multiclass with class weights is changed (it is now the same as in other `LogisticRegression` solvers with `'OvR'` strategy)

amueller commented:
why do we lose warm_start?

agramfort commented:
there is no reason not to support warm_start

I would put all sag related code in 1 file called sag.py (ie no sag_class.py)

agramfort commented:
travis is not happy

TomDLT commented:
For `warm_start`, how should I pass the option without adding parameters to `LogisticRegression`?

agramfort commented:
Ah... I thought there was a warm_start param in LR already. Keep it
for later then.

agramfort commented:
ping us when ready to merge.

thx

amueller commented:
I also thought there was a warm_start for LogisticRegressionCV... hum

amueller commented:
needs a rebase (probably for whatsnew)

TomDLT commented:
I implemented sag_ridge as a solver in `Ridge`, and changed accordingly some of the tests.

Currently, in order to match `Ridge` API, and compared to previous `SAGRegressor`, 
- `eta` is forced to `'auto'`
- we loose `random_state` and `warm_start`

agramfort commented:
loosing random_state is probably a bad idea. Any thought? Justify API
extension?

amueller commented:
yeah

amueller commented:
shouldn't LogisticRegression have a random_state for liblinear? Or is that only for hinge-loss?

agramfort commented:
Indeed it should

TomDLT commented:
No, `LogisticRegression` class has a `random_state`, and so has `sag_logistic`.
It is missing in `Ridge` class, so it is currently missing also in `sag_ridge`.

agramfort commented:
Ok i would add it then

TomDLT commented:
So I added `random_state` in `sag_ridge`.

TomDLT commented:
By the way, `_fit_liblinear` has a `random_state` parameter, but it is not used in `LogisticRegression` with `solver='liblinear'`.
Actually, `LogisticRegression` has a  `random_state` parameter that was never used (before SAG)

agramfort commented:
Oh boy ...

TomDLT commented:
Here is the implementation of SAG as 2 solvers (`sag_ridge` and `sag_logistic`).
I also updated the doc and the tests.

I checked speed performances, compared to previous class implementation(`SAGClassifier`):
Classifier for RCV1 dataset:
![classifier_rcv1_dloss](https://cloud.githubusercontent.com/assets/11065596/7841609/04d4947c-049e-11e5-94a6-2b3902fb9408.png)
Regressor for Alpha dataset:
![regressor_alpha_dloss_nointercept_alone](https://cloud.githubusercontent.com/assets/11065596/7841634/487e7314-049e-11e5-930e-a3152d335992.png)

TomDLT commented:
Please tell me what you think of the code.

Also, do we want to change the API for `Ridge` and `LogisticRegression`, in order to add a `warm_start` parameter?

amueller commented:
Why is there a change in speed? Shouldn't it be the same?

amueller commented on doc/modules/linear_model.rst:
```diff
@@ -696,6 +696,10 @@ be better calibrated than the default "one-vs-rest" setting.
 L-BFGS and newton-cg cannot optimize L1-penalized models, though,
 so the "multinomial" setting does not learn sparse models.
 
+The solver "sag" uses a Stochastic Average Gradient descent [3]_. It does not handle "multinomial" case, and is limited to L2-penalized models, yet it is
```
It uses ovr for multi-class, right?

TomDLT replied:
Yes

amueller replied:
Well, then I would say that ;)

agramfort replied:
line too long

amueller commented on doc/modules/linear_model.rst:
```diff
@@ -696,6 +696,10 @@ be better calibrated than the default "one-vs-rest" setting.
 so the "multinomial" setting does not learn sparse models.
 
+The solver "sag" uses a Stochastic Average Gradient descent [3]_. It does not handle "multinomial" case, and is limited to L2-penalized models, yet it is
+faster than other solvers for large datasets, when both the number of sample
```
samples

amueller commented on doc/modules/linear_model.rst:
```diff
@@ -696,6 +696,10 @@ be better calibrated than the default "one-vs-rest" setting.
 
+The solver "sag" uses a Stochastic Average Gradient descent [3]_. It does not handle "multinomial" case, and is limited to L2-penalized models, yet it is
+faster than other solvers for large datasets, when both the number of sample
+and the number of feature are large.
```
features

amueller commented on sklearn/utils/estimator_checks.py:
```diff
@@ -204,7 +205,7 @@ def set_fast_parameters(estimator):
         estimator.set_params(n_iter=5)
     if "max_iter" in params:
-        # NMF
+        warnings.simplefilter("ignore", ConvergenceWarning)
```
Why did you remove the nmf comment?

TomDLT replied:
@dsullivan7

dsullivan7 replied:
Because "max_iter" is now in the params of SAG, this no longer applies exclusively to NMF. I suppose it would have been wiser to instead make the comment "NMF or SAG".

amueller replied:
hum, max_iter is also in many other estimators, right? Not sure why I put NMF here then... well, never mind

amueller commented:
Does this implementation actually sample the data point to update randomly? I thought usually people just shuffle after each iteration. Have you compared the two approaches (or did anyone before)?
Do we want liblinear to be the default solver still?

amueller commented:
If you want to add warm_start, I'd be fine with that. it should also be supported for l-bfgs, right?

TomDLT commented:
>Does this implementation actually sample the data point to update randomly? 

Yes, each data point is chosen randomly, and the same data point can be chosen several times in a row ("draw with replacement").

>I thought usually people just shuffle after each iteration. Have you compared the two approaches (or did anyone before)?

In Mark Schmidt's [presentation](http://www.birs.ca/workshops/2014/14w5003/files/schmidt.pdf) (co-author of SAG), slide 76: 
"Does re-shuffling and doing full passes work better? NO !" 
It decreases speed performances. I did not tested it.

>Do we want 'liblinear' to be the default solver still?

The default solver is currently 'lbfgs'. I think we should keep it, since 'sag' solver is less efficient if the dataset is small.

TomDLT commented:
> If you want to add warm_start, I'd be fine with that. it should also be supported for l-bfgs, right?

Actually, there is two kinds of warm starting for SAG:
(1) reuse the `coef` as initial guess
(2) reuse the `coef` as initial guess, and store and reuse intermediate calculus [(SAG Paper, 4.4)] (http://arxiv.org/pdf/1309.2388v1.pdf) (better)

___
In `LogisticRegressionCV`, warm starting (1) is already used for SAG, LBFGS and Newton-CG.
I can just modify slightly in order to use warm starting (2) for SAG. (easy)

If we want warm starting in `LogisticRegression`, we only need to add a `warm_start` parameter, and warm starting (1) can be used for SAG, LBFGS and Newton-CG. If we want warm starting (2) for SAG, we need to store intermediate calculus, of size (n_samples + n_features).

___
In `Ridge`, other solvers do not handle warm starting. If we want warm starting (1) for SAG, we only need to add a `warm_start` parameter. If we want warm starting (2) for SAG, we need to store intermediate calculus, of size (n_samples + n_features).

`RidgeCV` currently does not handle SAG.

amueller commented:
What do the benchmarks on small datasets look like?

amueller commented:
I think doing (2) for SAG in LogisticRegressionCV would be great. Currently, using LogisticRegressionCV over doing GridSearchCV(LogisticRegression()) rarely gives any improvement in terms of speed.

GaelVaroquaux commented:
> Currently, using LogisticRegressionCV over doing
> GridSearchCV(LogisticRegression()) rarely gives any improvement in
> terms of speed.

On fMRI data it does :)

amueller commented:
And I thought we only add methods that are widely useful :P

GaelVaroquaux commented:
> And I thought we only add methods that are widely useful :P

Point taken. I think that for any very high dimensional learning problem
it helps. But I may be wrong.

agramfort commented on examples/linear_model/plot_sgd_comparison.py:
```diff
@@ -15,7 +15,7 @@
 
 from sklearn.cross_validation import train_test_split
-from sklearn.linear_model import SGDClassifier, Perceptron
+from sklearn.linear_model import SGDClassifier, Perceptron, SAGClassifier
```
cleanup

agramfort commented on sklearn/linear_model/__init__.py:
```diff
@@ -67,6 +68,8 @@
            'RidgeClassifier',
            'RidgeClassifierCV',
+           'SAGClassifier',
+           'SAGRegressor',
```
same here. SAGClassifier, SAGRegressor should disappear

StevenLOL commented:
Looks nice,  I can't wait to try this out.

TomDLT commented:
I implemented warm starting (1) in `LogisticRegression`, useful for solvers sag, lbfgs and newton-cg. It is used through an new parameter `warm_start`. 
I also added warm starting (2) in `LogisticRegressionCV` for sag solver (automatic warm starting behavior).
___
Benchmark of warm starting:
`LogisticRegressionCV(Cs=10, tol=0.01, solver="sag")` on RCV1 dataset :
- nothing : 296 seconds
- warm starting (1) : 257 seconds (-13%)
- warm starting (2) : 166 seconds (-44%)

___

>What do the benchmarks on small datasets look like?

Running 100 fits on iris dataset, 
class (1) vs (0 and 2):
![iris_1vr_100fits](https://cloud.githubusercontent.com/assets/11065596/7885941/16767dc8-0627-11e5-8a7f-375f2b0c9e50.png)
class (2) vs (0 and 1):
![iris_2vr_100fits](https://cloud.githubusercontent.com/assets/11065596/7885947/20a62a50-0627-11e5-8a43-6f893180a1c7.png)

I don't understand the difference for sag solver, yet I did not witness the same slow convergence with larger datasets. SAG is clearly interesting only for large datasets, so it is better to keep liblinear as default solver.

agramfort commented on examples/linear_model/plot_sgd_comparison.py:
```diff
@@ -31,11 +32,13 @@
                                                          C=1.0)),
     ("Passive-Aggressive II", PassiveAggressiveClassifier(loss='squared_hinge',
                                                           C=1.0)),
+    ("SAG", LogisticRegression(solver='sag', max_iter=5, C=1.e4 / X.shape[0]))
```
I get a convergence warning

```
/Users/alex/work/src/scikit-learn/sklearn/linear_model/sag.py:323: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
```

agramfort commented on sklearn/linear_model/logistic.py:
```diff
@@ -682,7 +688,7 @@ def logistic_regression_path(X, y, pos_class=None, Cs=10, fit_intercept=True,
             coefs.append(multi_w0)
         else:
-            coefs.append(w0)
+            coefs.append(np.copy(w0))
```
nitpick w0.copy() is more standard

agramfort commented on sklearn/linear_model/logistic.py:
```diff
@@ -1059,20 +1079,29 @@ def fit(self, X, y):
         self.coef_ = np.squeeze(self.coef_)
-        # For the binary case, this get squeezed to a 1-D array.
+        # For the binary case, this get squeezed to a 1-D array,
+        # or even 0-D if n_features == 1
```
really? X should always be 2d even if n_features==1

TomDLT replied:
Exactly, this is why I added lines 1102-1103

agramfort replied:
Sorry I don't understand how this can happen

TomDLT replied:
`logistic_regression_path` returns a list of array, and `LogisticRegression` uses a list to append the lists of array obtained for the different classes.
In order to transform the list of array into an array, we use `np.squeeze()`.

However, if we have only 1 class, it get squeezed to much, so we need to manually add another dimension. If we have only 1 class and 1 feature, it get squeezed to a scalar, so we need to add 2 other dimensions.
Actually, I wonder if we should not also add the case with 1 feature and multiclass.

TomDLT replied:
```Python
self.coef_ = np.squeeze(self.coef_)
if X.shape[1] == 1:
    self.coef_ = self.coef_[:, np.newaxis]
if n_classes == 1:
    self.coef_ = self.coef_[np.newaxis]
```
is better, it fixes the case with 1 feature and multiclass

agramfort replied:
Can you avoid the squeeze in the first place?

TomDLT replied:
Yes, actually, we can do
```Python
if self.multi_class == 'multinomial':
    self.coef_ = self.coef_[0]
else:
    self.coef_ = np.asarray(self.coef_)
```

agramfort replied:
Better

agramfort commented on sklearn/linear_model/ridge.py:
```diff
@@ -299,8 +311,9 @@ def ridge_regression(X, y, alpha, sample_weight=None, solver='auto',
-        X, y = _rescale_data(X, y, sample_weight)
+        if solver == 'cholesky':
+            # Sample weight can be implemented via a simple rescaling.
+            X, y = _rescale_data(X, y, sample_weight)
```
why only with cholesky?

TomDLT replied:
I misunderstood the doc:
> If sample_weight is set, then the solver will automatically be set to 'cholesky'

I guess it means: 
>If sample_weight is set and solver='auto', the solver will be set to 'cholesky'

which corresponds to actual behavior.

In the sparse case with `fit_intercept=True`, we cannot center the data, so the rescaling is not correct. As 'sag' solver handles correctly the sample weights, I did not rescale in this case.

I guess I can change it to `if solver != 'sag'`, to keep previous behavior for other solvers.
Previous behavior looks however suspicious to me in the sparse case with intercept (see issue #4710).

agramfort replied:
let's separate issues there and keep old behavior. But do add a comment to explain what you just said.
thx

agramfort commented on sklearn/linear_model/sag.py:
```diff
@@ -0,0 +1,327 @@
+    https://hal.inria.fr/hal-00860051/PDF/sag_journal.pdf
+
+    IMPORTANT NOTE: Ridge regressor with 'sag' solver and models from
+    linear_model in general depend on columns that are on the same scale.
```
"depend" is not clear. Say that it helps SAG to converge faster. It's not a requirement.

agramfort commented on sklearn/linear_model/sag.py:
```diff
@@ -0,0 +1,327 @@
+    Parameters
+    ----------
+    X : {array-like, sparse matrix}, shape (n_samples, n_features)
+            Training data
```
over indent. Fix below too.

agramfort commented on sklearn/linear_model/sag.py:
```diff
@@ -0,0 +1,327 @@
+
+def sag_logistic(X, y, coef_init, alpha=1e-4, sample_weight=None,
+                 max_iter=1000, tol=0.001, verbose=0, random_state=None,
+                 warm_start_mem=None):
```
the init and warm start params should be at the end. For example here logistic and ridge SAG have different API.

also why 2 warm start param? We should allow to have only 1 passed last.

agramfort commented on sklearn/linear_model/sag.py:
```diff
@@ -0,0 +1,327 @@
+    https://hal.inria.fr/hal-00860051/PDF/sag_journal.pdf
+
+    IMPORTANT NOTE: LinearRegression with 'sag' sovler and models from
+    linear_model in general depend on columns that are on the same scale.
```
same remark

agramfort commented on sklearn/linear_model/sag.py:
```diff
@@ -0,0 +1,327 @@
+    Parameters
+    ----------
+    X : {array-like, sparse matrix}, shape (n_samples, n_features)
+            Training data
```
idem

agramfort commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -665,3 +686,36 @@ def test_logreg_intercept_scaling_zero():
+                    if warm_start:
+                        assert_greater(2.0, cum_diff, msg)
+                    else:
+                        assert_greater(cum_diff, 2.0, msg)
```
tests are long and raise warnings:

```
.../Users/alex/work/src/scikit-learn/sklearn/linear_model/sag.py:169: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
..../Users/alex/work/src/scikit-learn/sklearn/linear_model/sag.py:323: ConvergenceWarning: The max_iter was reached which means the coef_ did not converge
  "the coef_ did not converge", ConvergenceWarning)
.......
----------------------------------------------------------------------
Ran 14 tests in 2.174s
```

TomDLT replied:
I removed the warnings, yet tests are long even in master branch.

agramfort commented on sklearn/linear_model/tests/test_ridge.py:
```diff
@@ -51,10 +51,10 @@ def test_ridge():
     # TODO: for this test to be robust, we should use a dataset instead
     # of np.random.
-    rng = np.random.RandomState(0)
+    rng = np.random.RandomState(1)
```
?

agramfort commented on sklearn/linear_model/tests/test_sag.py:
```diff
@@ -0,0 +1,693 @@
+import math
```
add author name + licence

agramfort commented on sklearn/linear_model/tests/test_sag.py:
```diff
@@ -0,0 +1,693 @@
+                              decimal=3)
+    assert_almost_equal(clf1.intercept_, spintercept1, decimal=1)
+
+    # TODO: uncomment when sparse Ridge with intercept will be fixed
```
this suggests that the other PR should be finished first.

agramfort commented on sklearn/tests/test_common.py:
```diff
@@ -160,7 +160,8 @@ def test_non_transformer_estimators_n_iter():
                              'RidgeClassifier', 'SVC', 'RandomizedLasso',
-                             'LogisticRegressionCV']):
+                             'LogisticRegressionCV', 'SAGClassifier',
+                             'SAGRegressor']):
```
cleanup

agramfort commented on sklearn/utils/seq_dataset.pxd:
```diff
@@ -25,10 +33,6 @@ cdef class ArrayDataset(SequentialDataset):
 
-    cdef void next(self, double **x_data_ptr, int **x_ind_ptr,
-                   int *nnz, double *y, double *sample_weight) nogil
-
```
why did you remove this?

TomDLT replied:
I don't know, is it required? @dsullivan7 ?
Do we also need to add on both `ArrayDataset` and `CSRDataset`
`cdef void _sample(self, double **x_data_ptr, int **x_ind_ptr, int *nnz, double *y, double *sample_weight, int current_index) nogil` ?

agramfort replied:
ping @dsullivan7

dsullivan7 replied:
I believe i added both next and random so that sgd could sample in order while sag could sample randomly. I'm away from the code right now, the tests for sgd still pass?

agramfort replied:
sgd still pass which bothers me...

agramfort commented:
any reason not to have warm_start for RidgeCV? don't we have this for CG solver?

TomDLT commented:
`RidgeCV` has two possibilities:
- if no CV generator is passed, it performs an efficient Generalized CV (Leave-One-Out), with SVD of Eigen decomposition:

> Compute eigendecomposition K = Q V Q^T.
    Then G = Q (V + alpha*Id)^-1 Q^T,
    where (V + alpha*Id) is diagonal.
    It is thus inexpensive to inverse for many alphas.

- if a CV generator is passed, if basically calls `GridSearchCV(Ridge(solver='auto'))`. Do we want to add a `solver` parameter?

Indeed, both sag and sparse_cg would benefit from a warm start. Do we want to add a `warm_start` parameter?

amueller commented:
This looks really good :)

amueller commented:
I think we should be as explicit as possible with advice on which solver to use, and in a fairly prominent place. So User Guide and Docstring? Maybe a short list / table in the userguide?

Small data: liblinear, multinomial: lbfgs/newton-cg, large data: sag, very large data: maybe SGDClassifier.
[Correct me if I'm wrong].

For LogisticRegressionCV, do we gain with SAG on small data? the warm-starts could help us.

amueller commented:
I wouldn't think it unreasonable to have a different default solver for LogisticRegressionCV than LogisticRegression if the benchmarks support it.

amueller commented:
Also maybe add sag to some of the the benchmarks in the bench folder, say mnist and covertype?

TomDLT commented:
**MNIST benchmark**
```
Classification performance:
===========================
Classifier               train-time   test-time   error-rate
------------------------------------------------------------
Nystroem-SVM                141.10s       3.12s       0.0228
ExtraTrees                   51.55s       0.46s       0.0288
RandomForest                 51.80s       0.40s       0.0304
SampledRBF-SVM              172.56s       1.87s       0.0488
LinearRegression-SAG         28.04s       0.05s       0.0815
CART                         24.52s       0.01s       0.1214
dummy                         0.00s       0.02s       0.8973
```
**Coverage benchmark**
```
Classification performance:
===========================
Classifier   train-time test-time error-rate
--------------------------------------------
liblinear     6.9757s    0.0122s     0.2305  
SGD           0.5395s    0.0122s     0.2315  
SAG           1.1922s    0.0170s     0.2345  
GaussianNB    0.3074s    0.0808s     0.4841 
```

amueller commented:
Nice. The PR should be [MRG], right?

agramfort commented:
yes. Changed the status. I still need to review one more time.

TomDLT commented:
[Lightning](https://github.com/mblondel/lightning)'s implementaton of SAG is a bit different from ours:
- it initializes the gradient sum by going through all the dataset once without updating the weights, whereas we follow [SAG 4.2](http://arxiv.org/pdf/1309.2388v1.pdf) guidelines
- there is no intercept fitting
- there is no automatic step size selection
- there are more losses available
- its stopping criterion is different. Ours is `max_change / max_weight < tol`, where max_change is the maximal change in the weights over n_sample iterations. In lightning, it is `violation / violation_init <  tol` where violation  is the L2 norm of the next step (without step size eta).

But more importantly, its just-in-time update seems incorrect to me in the sparse case (@mblondel). With RCV1 dataset, the two algorithms do not converge to the same minimum.

In the dense case, the minimum is the same and performances are similar: 
**Alpha dataset**
![alpha50k](https://cloud.githubusercontent.com/assets/11065596/8036660/9fc5a386-0df9-11e5-874e-93c902c7b8b4.png)
**iris OvR**
![iris_ovr](https://cloud.githubusercontent.com/assets/11065596/8036675/af05d7d0-0df9-11e5-87c6-5ccfaf478bd0.png)
**MNIST class1vR**
![mnist_1vr](https://cloud.githubusercontent.com/assets/11065596/8036682/b60584ea-0df9-11e5-91b4-2149f05b54c2.png)

I didn't find where the differences come from

mblondel commented:
SAG is sensitive to the step size. As you said, I didn't implement line search. What's wrong in the sparse case?

vene commented:
It's great that the minimum is the same and performance is so similar to `lightning`.
What happens when you feed both codebases the same dense dataset but converted to sparse, do they still go to different minima?

mblondel commented on sklearn/linear_model/ridge.py:
```diff
@@ -301,8 +312,13 @@ def ridge_regression(X, y, alpha, sample_weight=None, solver='auto',
+            # FIX: In the sparse case, with fit_intercept=True, we cannot
+            # center the data, so a rescaling is incorrect! See #4710.
+            # Sag solver can actually handle sample weights without rescaling,
+            # and can be used as a reference to solve #4710.
```
Could you write something more concise like: "SAG supports sample_weight directly. For other solvers, we implement sample_weight via a simple rescaling."

`fit_intercept=True` is wrong in the sparse case whether we use sample_weight or not. So I don't think the comment above is necessary.

For a temporary fix for #4710, maybe you can set SAG when solver="auto" and fit_intercept=True. Otherwise, raise an exception.

TomDLT replied:
Actually, as `fit_intercept` is not passed to `ridge_regression`, even sag solver cannot fit the intercept in the sparse case.
Shall I add a parameter to `ridge_regression` or always raise an error when `fit_intercept=True` in the sparse case?

mblondel commented:
Could you benchmark SAG on a regularization path with and without warm start? Curious to see the result.

BTW, iris is not a good dataset for benchmarking :)

TomDLT commented:
>What's wrong in the sparse case?

[This line](https://github.com/mblondel/lightning/blob/master/lightning/impl/sag_fast.pyx#L117) seems incorect to me.
One update for feature j:
`wj = K * (wj - eta * d / n / K)` with `K = 1 - eta * alpha`
Two updates for feature j:
``` Python
wj = K * (K * (wj - eta * d / n / K) - eta * d / n / K)
   = K * K (wj - eta * d / n / K - eta * d / n / K / K)
   = K * K (wj - eta * d / n * (1 / K + 1 / K / K))
  != K * K (wj - eta * d / n * 2 / K / K)
```
>What happens when you feed both codebases the same dense dataset but converted to sparse, do they still go to different minima?

Yes, exact same behavior than with dense version

TomDLT commented:
What do you mean by "on a regularization path" ?

vene commented:
I think @mblondel means to vary the alpha between 0 and the maximum value (that leads to all coefficients being essentially zero), to see what impact the strength of regularization has on the timing.

Like [this example](http://scikit-learn.org/stable/auto_examples/linear_model/plot_ridge_path.html) but plot the timing, I guess.

With warm start you can compute such a path very efficiently by initializing the solver with the solution to the stronger-regularized (?) problem that you just solved.

(?) or should it be in the other direction?

TomDLT commented:
In several part of Ridge docstring, we can read:
>Alpha corresponds to ``(2*C)^-1``

I think it corresponds to `C^-1`, as it lacks the factor in *both* the loss function and the regularization term.
My tests seems to confirm it. Do you agree?

mblondel commented:
> Do you agree?

For comparing ridge regression to ridge regression, you're correct, as both the loss and regularization terms are multiplied by 0.5.

> Alpha corresponds to ``C^-1`` in other linear models such as LogisticRegression or LinearSVC

Logistic regression and LinearSVC in liblinear don't multiply the the loss term per 0.5.
The verb "corresponds" feels inaccurate to me. Maybe "behaves similarly to" is better.

agramfort commented on sklearn/linear_model/logistic.py:
```diff
@@ -1331,7 +1370,7 @@ def fit(self, X, y):
                              "positive; got (tol=%r)" % self.tol)
 
-        X = check_array(X, accept_sparse='csr', dtype=np.float64)
+        X = check_array(X, accept_sparse='csr', dtype=np.float64, order='C')
```
should we order to C also for other solvers? I am not sure...

TomDLT replied:
In `LogisticRegression`, it was already order to C for other solvers:

https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/linear_model/logistic.py#L1033

agramfort commented on sklearn/linear_model/sag.py:
```diff
@@ -0,0 +1,342 @@
+
+    Returns
+    -------
+    coef_ : array, shape = [n_features]
```
shape = [n_features] -> shape (n_features,)

agramfort commented on sklearn/linear_model/sag_fast.pyx:
```diff
@@ -0,0 +1,411 @@
+        return p - y
+
+    def __reduce__(self):
+        return SquaredLoss, ()
```
is it a copy / paste from SGD code?

TomDLT replied:
Indeed, I will cimport it

agramfort commented on sklearn/linear_model/sag_fast.pyx:
```diff
@@ -0,0 +1,411 @@
+                #              &xnnz,
+                #              &y,
+                #              &sample_weight)
+                # current_index = itr
```
cleanup

agramfort commented on sklearn/linear_model/sag_fast.pyx:
```diff
@@ -0,0 +1,411 @@
+                          " MinMaxScaler might help.") % (total_iter + 1))
+
+    return intercept, num_seen, max_iter_reached, intercept_sum_gradient
+
```
2 lines between functions

agramfort commented on sklearn/linear_model/sag_fast.pyx:
```diff
@@ -0,0 +1,411 @@
+        feature_hist[j] = (total_iter + 1) % n_samples
+
+    cumulative_sums[total_iter % n_samples] = 0.0
+
```
2 lines between functions

agramfort commented on sklearn/linear_model/sag_fast.pyx:
```diff
@@ -0,0 +1,411 @@
+    else:
+        # Lipschitz for squared loss
+        return 1.0 / (max_squared_sum + fit_intercept + alpha)
+
```
idem

agramfort commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -245,20 +247,20 @@ def test_consistency_path():
-    for method in ('lbfgs', 'newton-cg', 'liblinear'):
+    for method in ('lbfgs', 'newton-cg', 'liblinear', 'sag'):
         coefs, Cs = f(logistic_regression_path)(
-            X, y, Cs=Cs, fit_intercept=False, tol=1e-16, solver=method)
```
you're changing the default tol params.

is that on purpose?

@MechCoder any idea?

TomDLT replied:
It is just to speed up the test, from `0.545s` to `0.155s`.
The tol can be lowered to 1e-3 if we accept to check only 3 decimal instead of 4. The test would last `0.102s`, but I don't want to decrease to much not to affect stability.

Without 'sag', the tests used to last `0.501s`

ogrisel replied:
0.155s is good enough.

agramfort replied:
+1 for RCV1 dataset loader.

agramfort commented on sklearn/linear_model/tests/test_logistic.py:
```diff
@@ -488,29 +490,44 @@ def test_ovr_multinomial_iris():
     X, y = make_classification(n_features=10, n_informative=5, random_state=0)
-    clf_n = LogisticRegression(solver='newton-cg', fit_intercept=False)
-    clf_n.fit(X, y)
+    clf_new = LogisticRegression(solver='newton-cg', fit_intercept=False)
```
clf_new ?

agramfort replied:
ok new means newton...

ogrisel replied:
Better call it `clf_newton`, otherwise it's confusing.

agramfort commented:
besides LGTM

nice work @TomDLT !

you'll need to rebase.

agramfort commented:
+1 for merge on my side.

TomDLT commented:
rebased
another review?

ogrisel commented:
I have enabled the appveyor continuous integration and it reports a build issue under windows with the MSVC compiler (both with MSVC 2008 for Python 2.7 and MSVC 2010 for Python 3.4). Here are the details. Apparently the is a problem with the `fmax` symbol:

```
cl.exe /c /nologo /Ox /MD /W3 /GS- /DNDEBUG -IC:\Python34-x64\lib\site-packages\numpy\core\include -IC:\Python34-x64\lib\site-packages\numpy\core\include -IC:\Python34-x64\include -IC:\Python34-x64\include /Tcsklearn\linear_model\sag_fast.c /Fobuild\temp.win-amd64-3.4\Release\sklearn\linear_model\sag_fast.obj
link.exe /DLL /nologo /INCREMENTAL:NO /LIBPATH:C:\Python34-x64\libs /LIBPATH:C:\Python34-x64\PCbuild\amd64 /LIBPATH:build\temp.win-amd64-3.4 /EXPORT:PyInit_sag_fast build\temp.win-amd64-3.4\Release\sklearn\linear_model\sag_fast.obj /OUT:build\lib.win-amd64-3.4\sklearn\linear_model\sag_fast.pyd /IMPLIB:build\temp.win-amd64-3.4\Release\sklearn\linear_model\sag_fast.lib /MANIFESTFILE:build\temp.win-amd64-3.4\Release\sklearn\linear_model\sag_fast.pyd.manifest
sag_fast.obj : warning LNK4197: export 'PyInit_sag_fast' specified multiple times; using first specification
   Creating library build\temp.win-amd64-3.4\Release\sklearn\linear_model\sag_fast.lib and object build\temp.win-amd64-3.4\Release\sklearn\linear_model\sag_fast.exp
sag_fast.obj : error LNK2019: unresolved external symbol fmax referenced in function __pyx_pf_7sklearn_12linear_model_8sag_fast_sag_sparse
build\lib.win-amd64-3.4\sklearn\linear_model\sag_fast.pyd : fatal error LNK1120: 1 unresolved externals
error: Command "link.exe /DLL /nologo /INCREMENTAL:NO /LIBPATH:C:\Python34-x64\libs /LIBPATH:C:\Python34-x64\PCbuild\amd64 /LIBPATH:build\temp.win-amd64-3.4 /EXPORT:PyInit_sag_fast build\temp.win-amd64-3.4\Release\sklearn\linear_model\sag_fast.obj /OUT:build\lib.win-amd64-3.4\sklearn\linear_model\sag_fast.pyd /IMPLIB:build\temp.win-amd64-3.4\Release\sklearn\linear_model\sag_fast.lib /MANIFESTFILE:build\temp.win-amd64-3.4\Release\sklearn\linear_model\sag_fast.pyd.manifest" failed with exit status 1120
Command exited with code 1
```

ogrisel commented on doc/modules/sgd.rst:
````diff
@@ -192,7 +197,11 @@ specified via the parameter ``epsilon``. This parameter depends on the
-Averaging can be enabled by setting ```average=True```
+Averaging can be enabled by setting ```average=True```.
+
+For regression with a squared loss, another variant of SGD with an averaging
````
For regression with a squared loss and a l2 penalty, another...

ogrisel commented:
For the `fmax` symbol I guess we will have to redefine it as we did in `cd_fast.pyx` instead of relying on the `libc`.

ogrisel commented:
Could you please adapt the convergence plots of https://github.com/scikit-learn/scikit-learn/pull/4738#issuecomment-103510464  into a new `benchmark/bench_20newsgroups_logreg_convergence.py`?

And maybe another similar script for low dimensional dense data such as covertype.

ogrisel commented:
I don't want to delay this PR further but out of curiosity how much work would it require to add support for the multinomial loss + SAG solver combo? If no major blocker identified it would be great to have that as well (probably in a later PR).

ogrisel commented on sklearn/linear_model/sag.py:
```diff
@@ -0,0 +1,349 @@
+    parameters towards the zero vector using the squared euclidean norm
+    L2.
+
+    This implementation works with data represented as dense or sparse numpy
```
a dense numpy array or a scipy sparse matrix

ogrisel commented:
Actually about the benchmarks, it would be great to have an RCV1 dataset loader in scikit-learn (also for another PR).

vene commented on doc/modules/sgd.rst:
````diff
@@ -192,7 +197,11 @@ specified via the parameter ``epsilon``. This parameter depends on the
-Averaging can be enabled by setting ```average=True```
+Averaging can be enabled by setting ```average=True```.
+
+For regression with a squared loss and a l2 pernalty, another variant of
````
typo "penalty"

vene commented on sklearn/linear_model/sag.py:
```diff
@@ -0,0 +1,349 @@
+    a constant learning rate. The inspiration for SAG comes from "Minimizing
+    Finite Sums with the Stochastic Average Gradient" by Mark Schmidt,
+    Nicolas Le Roux, and Francis Bach. 2013. <hal-00860051>
+    https://hal.inria.fr/hal-00860051/PDF/sag_journal.pdf
```
This could go in a References section I guess.

ogrisel replied:
+1

TomDLT commented:
> out of curiosity how much work would it require to add support for the multinomial loss + SAG solver combo?

I think it's feasible. What do you think @agramfort ?

> Could you please adapt the convergence plots into a new benchmark?

I will. Maybe in another PR, after RCV1 loader is merged? #4908 
___
Thanks for the reviews, I am working on AppVeyor failure.

agramfort commented:
plan:

- merge RCV1 fetcher
- add benchmark script based on RCV1 to this PR
- merge this PR
- do multinomial in another PR later.

ogrisel commented on sklearn/linear_model/sag.py:
```diff
@@ -0,0 +1,355 @@
+MAX_INT = np.iinfo(np.int32).max
+
+"""For sparse data intercept updates are scaled by this decay factor to avoid
+intercept oscillation."""
```
Is there a reference for this magic number? Maybe mention the paragraph of the paper where this is documented.

Also: please use the `#` sign for inline comments. `"""` should be reserved for docstrings. Is there.

ogrisel commented on sklearn/linear_model/sag_fast.pyx:
```diff
@@ -0,0 +1,325 @@
+    return y
+
+
+def sag_sparse(SequentialDataset dataset,
```
why is this suffixed with `_sparse`? It works both for sparse and dense inputs right? I think this name is confusing if we don't implement a `sag_dense` counterpart.

dsullivan7 replied:
This is an artifact of me implementing both standard and sparse implementations in the pyx file. Because I removed the standard implementation, it's fine to remove the ```_sparse``` implementation prefix.

ogrisel commented on sklearn/linear_model/tests/test_sag.py:
```diff
@@ -0,0 +1,710 @@
+    fit_intercept = True
+    eta = get_eta(X, alpha, fit_intercept)
+    clf = LogisticRegression(solver="sag", fit_intercept=fit_intercept,
+                             tol=.00000000001, C=1. / alpha / n_samples,
```
It's more readable to use `tol=1e-12` here.

ogrisel commented on sklearn/linear_model/tests/test_sag.py:
```diff
@@ -0,0 +1,710 @@
+    assert_array_almost_equal(weights, clf.coef_, decimal=10)
+    assert_array_almost_equal(intercept, clf.intercept_, decimal=10)
+    assert_array_almost_equal(weights2, clf.coef_, decimal=10)
+    assert_array_almost_equal(intercept2, clf.intercept_, decimal=10)
```
I don't really see the point of comparing our cython implementation of sag to 2 pure python implementations maintained in a test file. As the problem is convex, the minimum is global, we could compare the coef of a fitted with `LogisticRegression(solver="sag", tol=1e-12)` with the coefs of `LogisticRegression(solver="lbfgs", tol=1e-12)`  and `LogisticRegression(solver="newton-cg", tol=1e-12)` instead.

We would not have to maintain the pure Python versions of SAG in the tests. What do other people think?

dsullivan7 replied:
I used the pure python implementation to more precisely measure the correctness of the algorithm, to ensure that the results I got were not coincidental. I found that it's easier to read the pure python implementation rather than dig into the pyx sparse implementation. I don't think it adds that much overhead and is a fairly good sanity checkpoint.

ogrisel replied:
I agree it might have been useful as a scaffold when initially implementing and debugging the Cython version but now that we have other convergence tests to detect regression I am not sure we should keep the 2 Python variants in the tests from a maintenance point of view.

What do other people think (@GaelVaroquaux @agramfort @TomDLT ) ?

ogrisel commented on sklearn/linear_model/tests/test_sag.py:
```diff
@@ -0,0 +1,710 @@
+    n_iter = 50
+    fit_intercept = False
+    np.random.seed(10)
+    X = np.random.random((n_samples, n_features))
```
Please do not mutate the singleton RNG but rather use an isolated instance and avoid causing hard to understand side-effects in the test execution.

ogrisel commented:
As we release the GIL in the Cython sag solver, we can leverage the threading backed of joblib for that solver.

In the fit method of `LogisticRegressionCV` we can do:

```python
        # The SAG solver releases the GIL so it's more efficient to use
        # threads for this solver.
        backend = 'threading' if self.solver == 'sag' else 'multiprocessing'
        fold_coefs_ = Parallel(n_jobs=self.n_jobs, verbose=self.verbose,
                                  baclend=backend)(
            path_func(X, y, train, test, pos_class=label, Cs=self.Cs,
                      fit_intercept=self.fit_intercept, penalty=self.penalty,
                      dual=self.dual, solver=self.solver, tol=self.tol,
                      max_iter=self.max_iter, verbose=self.verbose,
                      class_weight=self.class_weight, scoring=self.scoring,
                      multi_class=self.multi_class,
                      intercept_scaling=self.intercept_scaling
                      )
            for label in iter_labels
            for train, test in folds)
```

Also the in the `fit` method of the `LogisticRegression` class, the following loop could be similarly joblib'ed:

```python
        for ind, class_ in enumerate(classes_):
            coef_, _ = logistic_regression_path(
                X, y, pos_class=class_, Cs=[self.C],
                fit_intercept=self.fit_intercept, tol=self.tol,
                verbose=self.verbose, solver=self.solver,
                multi_class=self.multi_class, max_iter=self.max_iter,
                class_weight=self.class_weight)
            self.coef_.append(coef_[0])
```

TomDLT committed:
BENCH add rcv1 benchmark

TomDLT committed:
FIX bug and add threading

TomDLT commented:
@ogrisel 
I finally understood the bug: in [this line](https://github.com/scikit-learn/scikit-learn/pull/4738/files#diff-daba6a5cb7eb568260e4c88d83bf419eR23), if y is not already np.float64, a copy is done, then passed to the Dataset, then discarded. The Dataset only keeps a pointer, so the memory is then free and the data can be easily corrupted.

mblondel commented:
@TomDLT One solution is to keep a reference to the data in the dataset object, i.e., add `self.X = X` in
https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/utils/seq_dataset.pyx#L82

This way the garbage collector can't release the data.

TomDLT commented:
Update:
- A new benchmark on RCV1 dataset compares SAG with other logistic regression solvers.
- `SequentialDataset` now has some python testing function.
- The bug on windows in solved: more checking is done on dtype and a reference to the data is now kept on `SequentialDataset` to avoid garbage collection.
- `LogisticRegression` now handles parallel computing for OvR multiclass. SAG also handles multi-threading for OvR multiclass and for `LogisticRegressionCV`.

___
- I tried to reuse the random generator of `_tree.pyx` in `cd_fast.pyx` and `seq_dataset.pyx`, instead of duplicating code. Travis is happy but AppVeyor isn't. I am looking into it.

GaelVaroquaux commented:
>   • A new benchmark on RCV1 dataset compares SAG with other logistic regression
>     solvers.

Can we have a screenshot of the results in the PR, please.

Can you also compare to dual coordinate ascent, in lightning? @mblondel
seemed to think that it would be fast.

TomDLT commented:
**Benchmark on rcv1 (sparse 781265 x 47236):**
![rcv1](https://cloud.githubusercontent.com/assets/11065596/8805614/7a3b2cd0-2fd2-11e5-8b63-5631706b1348.png)

As Lightning SAG seems incorrect to me in the sparse case, I ran the benchmark on a dense small subset of rcv1:
**Benchmark on rcv1 subset (dense 781265 x 200):**
![rcv1_200](https://cloud.githubusercontent.com/assets/11065596/8805695/fa88b22c-2fd2-11e5-92e2-522a6cdc596c.png)
Curiously the minimum is not exactly the same, I couldn't find why.

The Dual Coordinate Descent (in lightning) does not handle the Logistic loss, so I couldn't add it to this benchmark.

agramfort commented:
if you compare to liblinear don't fit the intercept as it is penalized.
Otherwise set intercept_scaling to 1e4 or something. Can you test again?

TomDLT committed:
ENH add n_iter in logistic regression

GaelVaroquaux commented:
Thanks a lot. This is very interesting. And congratulations for the awesome work.

ogrisel commented:
@TomDLT can you include the plots for the train and validation scores in this thread? Just to show the others that in practice good solvers are useless from a learning point of view on RCV1 ;)

ogrisel commented:
@TomDLT can you please run a quick benchmark on `LogisticRegressionCV` with `n_jobs in [1, 2, 4, 8]` and `solver='sag'` on a multiclass classification problem such as a subset of mnist for instance to measure the scalability of the threading-based parallelism on a 8 cores+ host?

TomDLT commented:
**Benchmark on rcv1 (sparse 781265 x 47236):**
Previous plots showed `log10(train_loss - min_train_loss)`, which shows the speed of convergence. 
As suggested by @ogrisel, we can also plot the `score` on training and testing sets, which is what matter from a learning point of view (the goal is not to minimize a loss function, but rather to find a good classifier). In the case of the RCV1 dataset, all solvers are quite equivalent in term of score, as we can see in following plots:
**Score on training dataset (781265 samples):**
![rcv1_train_score](https://cloud.githubusercontent.com/assets/11065596/8821726/33ff8b7e-3060-11e5-9a8a-7a12c361f419.png)
**Score on testing dataset (23149 samples):**
![rcv1_test_score](https://cloud.githubusercontent.com/assets/11065596/8821729/36138b54-3060-11e5-8d16-b2b27044ed35.png)

Note for example the SGD case: the convergence is quite slow, but it is in the top classifiers in term of both training and testing scores. 
**Conclusion:** This benchmark still makes sense to show speed of convergence of different solvers. However, for some datasets like RCV1, many solvers are equivalent to find quickly a good classifier. For some other datasets, it might not be the case.
___
As noted by @agramfort, we can see in my previous post that the liblinear cases do not converge to the same minimum, because it regularizes the intercept (see how they do not go below 1e-10). To reduce the intercept regularization, we can increase the intercept scaling.
**With `intercept_scaling=1e4`**
![rcv1_intercept_scaling1e4](https://cloud.githubusercontent.com/assets/11065596/8821736/3e4134de-3060-11e5-88ba-6b44ac937752.png)
Here the liblinear solvers are not very good: the convergence is delayed by a too strong intercept_scaling.

**Without fitting the intercept:**
![rcv1_no_intercept](https://cloud.githubusercontent.com/assets/11065596/8821738/405206e0-3060-11e5-9dd6-c55205d8b8f4.png)
Here the liblinear solvers compete with the other, especially in the dual case.

**Conclusion:** The liblinear solvers are very efficient. However, as the intercept is regularized, it is difficult to compare to other solver while fitting the intercept, even with different values for `intercept_scaling`.

TomDLT commented:
> @TomDLT can you please run a quick benchmark on `LogisticRegressionCV` with `n_jobs in [1, 2, 4, 8]` and `solver='sag'` on a multiclass classification problem such as a subset of mnist for instance to measure the scalability of the threading-based parallelism on a 8 cores+ host?

On a 6 cores CPU:
```
n_samples : [691, 788, 699, 714, 682, 632, 687, 729, 684, 694]
LogisticRegression OvR multiclass:
n_jobs = 1 - time = 25.5 sec
n_jobs = 2 - time = 13.3 sec
n_jobs = 4 - time = 8.0 sec
n_jobs = 6 - time = 5.7 sec

n_samples : [593, 674, 596, 0, 0, 0, 0, 0, 0, 0]
LogisticRegressionCV:
n_jobs = 1 - time = 41.5 sec
n_jobs = 2 - time = 24.1 sec
n_jobs = 4 - time = 15.3 sec
n_jobs = 6 - time = 11.8 sec
```

See [this gist](https://gist.github.com/TomDLT/d717bf372f944f8ebc0b) for the code

ogrisel commented:
I tried on a machine with 12 cores, the scalability looks nice:

```
n_samples : [691, 788, 699, 714, 682, 632, 687, 729, 684, 694]
LogisticRegression OvR multiclass:
n_jobs = 1 - time = 21.4 sec
n_jobs = 2 - time = 11.3 sec
n_jobs = 4 - time = 6.7 sec
n_jobs = 8 - time = 4.7 sec
n_jobs = 10 - time = 2.9 sec
n_samples : [593, 674, 596, 0, 0, 0, 0, 0, 0, 0]
LogisticRegressionCV:
n_jobs = 1 - time = 22.9 sec
n_jobs = 2 - time = 15.6 sec
n_jobs = 4 - time = 11.5 sec
n_jobs = 8 - time = 9.7 sec
n_jobs = 12 - time = 7.9 sec
```

I had to edit the script to set `max_iter=1000, tol=1e-2` for the `LogisticRegression` OvR task and `max_iter=1000, tol=1e-3` for the `LogisticRegressionCV` task otherwise I would get convergence warnings.

What you could do to improve the scalability further is to add `check_input=True` kwarg to `logistic_regression_path`, `sag_logistic` and `sag_ridge` and maybe other inner functions to be able to do the input data check once in the `fit` method of `LogisticRegression` and `LogisticRegressionCV` and then disable the nested redundant calls to `check_array` by passing `check_input=False` in the joblib parallel calls to those functions.

ogrisel commented on sklearn/linear_model/sag.py:
```diff
@@ -0,0 +1,354 @@
+
+def make_dataset(X, y, sample_weight, random_state):
+    # check which type of Sequential Dataset is needed
+    seed = random_state.randint(1, np.iinfo(np.int32).max)
```
Please add an inline comment to explain why the `seed` should not be zero here. Please also feel free to open an issue to do an audit of the proper seeding of all occurrences of `our_rand_r` in the code base.

TomDLT replied:
what about `self.seed = max(1, seed)` in SequantialDataset `__cinit__` ?

ogrisel replied:
1 will be twice more likely than any other value instead of being uniform. I don't think this is that of a problem a problem in practice. Anyway we still need to put inline comments to make it explicit that `out_rand_r` should never be seeded with 0 and make an audit of our code to prevent that bug from happening.

ogrisel commented on sklearn/linear_model/sag_fast.pyx:
```diff
@@ -0,0 +1,325 @@
+    # helper variable for indexes
+    cdef int idx
+    # the total number of interations through the data
+    cdef int total_iter = 0
```
This variable name is misleading. It should better be named `n_updates` and the above inline comment:

```
# the total number of individual weight updates applied to the model
```

ogrisel commented on sklearn/linear_model/sag_fast.pyx:
```diff
@@ -0,0 +1,325 @@
+                          " with StandardScaler or"
+                          " MinMaxScaler might help.") % (total_iter + 1))
+
+    return intercept, num_seen, max_iter_reached, intercept_sum_gradient
```
Instead of returning `max_iter_reached`, we should return `n_iter = max(n_updates // n_samples, 1)` and store it as an `n_iter_` attribute on `LogisticRegression` instance. Do a `git grep "self.n_iter_ ="` to see other occurrences of this pattern in the code base.

The convergence warning should be raised  by the caller if `n_iter == max_iter`.

TomDLT replied:
About `n_iter = max(n_updates // n_samples, 1)` : 
If `max_iter=1`, and the solver stops before the first full pass, we don't want to have a convergence warning. Is it ok to have in this case `self.n_iter_ = 0` ?

ogrisel replied:
This is not possible as we only do convergence checks at the end of a full pass over the training samples.

ogrisel replied:
actually we don't need, the `max`, we can just have `n_iter = n_updates // n_samples`

ogrisel replied:
Actually it would even be better to have a variable `n_iter = 0` initialized at the beginning instead of `total_iter / n_updates` and then increment it by one after each epoch (instead of each update). The we never need to divide it by `n_samples`.

ogrisel commented on sklearn/linear_model/sag.py:
```diff
@@ -0,0 +1,354 @@
+    if max_iter is None:
+        max_iter = 1000
+
+    y = check_array(y, dtype=np.float64, ensure_2d=False)
```
You should check `X` here as well, unless `check_input is False`.

ogrisel commented on sklearn/linear_model/sag_fast.pyx:
```diff
@@ -0,0 +1,325 @@
+        raise ValueError(("Floating-point under-/overflow occurred at epoch"
+                          " #%d. Lowering the eta0 or scaling the input data"
+                          " with StandardScaler or"
+                          " MinMaxScaler might help.") % (total_iter + 1))
```
This is wrong: `total_iter` is not the number of passes over the data.

ogrisel commented on sklearn/utils/tests/test_seq_dataset.py:
```diff
@@ -0,0 +1,73 @@
+        _, _, _, idx1 = dataset1._next_py()
+        _, _, _, idx2 = dataset2._next_py()
+        assert_equal(idx1, i)
+        assert_equal(idx2, i)
```
you can also test `assert_equal(idx1, idx2)` with `_random_py` here.

TomDLT committed:
ENH add check_input parameter

TomDLT commented:
@ogrisel I added a `check_input` parameter to avoid multiple check.

I also changed `logistic_regression_path` to avoid modification of `y` along the path. What do you think about the `copy` parameter? Shall we deprecate it?

TomDLT committed:
deprecate 'copy' in logistic_regression_path

ogrisel commented:
> What do you think about the copy parameter? Shall we deprecate it?

It's useless now, +1 for deprecation if it was already there in 0.16.1. If it was introduced after 0.16.1, then you can just remove it.

ogrisel commented:
Skipping the redundant input check does not seem to bring much of a gain but still think it's a good idea anyway.

+1 for merge once the removal / deprecation of copy is done.

ogrisel commented on benchmarks/bench_rcv1_logreg_convergence.py:
```diff
@@ -0,0 +1,234 @@
+#     nnz_in_each_features = X_csc.indptr[1:] - X_csc.indptr[:-1]
+#     X = X_csc[:, np.argsort(nnz_in_each_features)[-200:]]
+#     X = X.toarray()
+#     print("dataset: %.3f MB" % (X.nbytes / 1e6))
```
Why is this commented out?

TomDLT replied:
I ran the benchmark with `fit_intercept=False` and full sparse dataset, and forgot to uncomment it.
I will

ogrisel commented:
Actually there is another optim: please pre-compute the step size once `get_auto_eta` in fit (it's independent of the y so it should be the same for all binary classifier) and pass it as optional argument to avoid re-computing it each time.

TomDLT commented:
`copy` in `logistic_regression_path` has been added in f9871cd824727ebad95bdc01e29f41329f3355a6, so it does exist in version 16.1.
I will deprecate it.

TomDLT committed:
ENH avoid recomputing max_squared_sum in get_auto_step_size for SAG solver

TomDLT commented:
Update:
- `copy` in `logistic_regression_path` is now deprecated
- the automatic step size selection in SAG is decomposed into `get_max_squared_sum` which goes through all the samples (slow) and `get_auto_step_size` which is fast. In LogisticRegression, for OvR multiclass and for cross-validation,  `max_squared_sum` is now computed only once.

TomDLT committed:
ENH add n_iter in ridge

amueller commented:
Is this waiting for reviews or anything else to do?

TomDLT commented:
Yes, this is ready for more reviews.
I am working on adding multinomial logistic regression with SAG, but it will be another PR.
Test failures are unrelated.

ogrisel commented on sklearn/linear_model/ridge.py:
```diff
@@ -351,14 +395,15 @@ def ridge_regression(X, y, alpha, sample_weight=None, solver='auto',
         # When y was passed as a 1d-array, we flatten the coefficients.
         coef = coef.ravel()
 
-    return coef
```
The `ridge_regression` function is part of the public API so we cannot change the values returned by default. Please add a new kwarg named `return_n_iter=False` to make it explicit to get this behaviour. Also please update the docstring of this function to document the new return value and the new matching parameter.

TomDLT replied:
Done

ogrisel commented:
Sorry for the slow response, this PR needs a rebase on top of the current master.

ogrisel commented on doc/modules/linear_model.rst:
```diff
@@ -697,9 +692,25 @@ Setting `multi_class` to "multinomial" with the "lbfgs" or "newton-cg" solver
 regression model, which means that its probability estimates should
 be better calibrated than the default "one-vs-rest" setting.
-L-BFGS and newton-cg cannot optimize L1-penalized models, though,
+"lbfgs" and "newton-cg" solvers cannot optimize L1-penalized models, though,
```
This is also the case for the `"sag"` solver.

ogrisel commented on doc/modules/linear_model.rst:
```diff
@@ -697,9 +692,25 @@ Setting `multi_class` to "multinomial" with the "lbfgs" or "newton-cg" solver
 
+The solver "sag" uses a Stochastic Average Gradient descent [3]_. It does not
+handle "multinomial" case, and is limited to L2-penalized models, yet it is
+faster than other solvers for large datasets, when both the number of samples
```
I would rather say "often faster".

ogrisel commented on sklearn/linear_model/logistic.py:
```diff
@@ -914,15 +949,21 @@ class LogisticRegression(BaseEstimator, LinearClassifierMixin,
+        - For multiclass problems, only 'newton-cg' and 'lbfgs' handle
+            multinomial loss; 'sag' and 'liblinear' are limited to
+            one-versus-rest schemes.
+        - 'newton-cg', 'lbfgs' and 'sag' only handle L2 penalty.
```
I would add a note to state that `'sag'` fast convergence is only guaranteed on features with approximately the same scale, possibly by preprocessing the data with a scaler from `sklearn,preprocessing`.

A similar remark would also hold for the docstring of the `Ridge` class.

TomDLT replied:
Done

ogrisel commented:
I made another quick pass over this PR. +1 for merge on my side once my final comments are addressed.

amueller commented:
After this is merged, do we maybe want to add an "auto" algorithm that chooses between liblinear and sag and l-bfgs? Having interdependent options like (this)[http://stackoverflow.com/questions/32074630/training-logistic-regression-using-scikit-learn-for-multi-class-classification] is not nice, and for large datasets, maybe we want to default to sag?

fabianp commented on sklearn/linear_model/sag_fast.pyx:
```diff
@@ -0,0 +1,341 @@
+    cumulative_sums[itr % n_samples] = 0.0
+
+
+cdef double dot(double* x_data_ptr, int* x_ind_ptr, double* w_data_ptr,
```
should use the BLAS ddot function instead of defining another dot product function. Look inside linear_model/cd_fast.pyx or cluster/_k_means.pyx for example usage.

dsullivan7 replied:
I believe this will be faster for sparse implementations. It skips adding 0 products.

fabianp replied:
OK I had not realized that you pass an array of indices. Maybe rename to sparse_dot or add a docstring ?

fabianp commented on sklearn/linear_model/sag.py:
```diff
@@ -0,0 +1,394 @@
+    return coef_, n_iter
+
+
+def sag_logistic(X, y, sample_weight=None, alpha=1e-4, max_iter=1000,
```
Unless there's a good reason, I think alpha be equal to 1 since that is the default of LogisticRegression(). Same for sag_ridge

ogrisel replied:
`alpha = 1 / (n_samples * C)` right, so if `C=1` by default at the estimator level, it's not equivalent to `alpha=1` here.

fabianp replied:
I'm missing something since I don't see why `alpha = 1 / (n_samples * C)`. 

Comparing `alpha` from [here](http://scikit-learn.org/stable/modules/linear_model.html#ridge-regression) with `C` from [here](http://scikit-learn.org/stable/modules/linear_model.html#logistic-regression) it to be that way.

Also, logistic_regression_path calls sag_logistic with alpha=1. / C (logistic.py#L696)

agramfort replied:
checks have been done on cost function convergence so I am not sure. But It
is likely to be correct

ogrisel replied:
> Also, logistic_regression_path calls sag_logistic with alpha=1. / C (logistic.py#L696)

Indeed. I was confused with the parameterization in `SGDClassifier` where I think `alpha = 1 / (C * n_samples)`. To avoid confusion I think we should only use the `C` parameterization in the SAG code. WDYT?

agramfort replied:
I have no strong feeling here.

fabianp replied:
+1 on that. It is sufficiently confusing that some models accept alpha and others C as to have to deal with this distinction within the same model. To keep that consistency I would use `C` for logistic methods (sag_logistic) and `alpha` for ridge methods (sag_ridge).

fabianp commented:
Besides the minor comments of before, the code looks good to me and I'm very happy to have this in scikit-learn

fabianp commented:
@TomDLT : I don't know how involved you are still on this pull request. If you want I can take it over from here (respecting of course the authorship of the commits).

ogrisel commented:
I believe @TomDLT is on vacation at the moment. He should be back online soon :)

agramfort commented:
he'll be back on Sept 7. Send him a PR to his branch while you're at it.

fabianp commented:
some more feedback (I start using it and some some usability issues come to mind). 

The function signature of `sag_ridge` and  `sag_logistic` is quite similar but its output is very different, which feels strange. The first one returns `coef_, n_iter` while the second one returns a dict with many keys (including coef) and n_iter. I understand that sag_logistic might want to return more information, but to make it more consistent I would make it such that both return a triplet `(coef_, n_iter, warm_start_mem)` where the last one contains the information in warm_start_mem from sag_logistic except for coef_. What do you think?

agramfort commented:
I like that.

ogrisel commented:
+1 as well for consistency.

TomDLT committed:
refactor sag_logistic and sag_ridge in sag_solver

TomDLT committed:
alpha clarification

TomDLT commented:
Thanks a lot for all your reviews !
___

Changing the signature of `sag_logistic` and `sag_ridge` would have lead to extremely similar functions, so I decided to merge them in one function `sag_solver`. 
___
About default alpha value, I changed it to 1, to be consistent with Ridge's default alpha, and LogisticRegression's default C.
___
About the scaling of alpha by `n_samples`, the confusion comes from :
`alpha_ridge = 1. / C_logistic`
`alpha_sgd = 1. / C_logistic / n_samples`

For example in this [benchmark](https://github.com/scikit-learn/scikit-learn/pull/4738/files#diff-174d08d6b044dde709b6ec174854cc93R190), I need to scale `alpha_sgd` to obtain the same minimum.

`alpha_sag` was initially equal to `alpha_sgd`.
I chose to change it to `alpha_sag = alpha_ridge = 1. / C_logistic`, and to scale it by `n_samples` internally.
Choosing `C_sag = C_logistic = 1. / alpha_ridge` would be identical.

To avoid confusion, I now changed the named in `alpha_scaled` internally.
___
As suggested, I also added a boolean parameter `return_n_iter` in `ridge_regression` public function to avoid changing its signature.

fabianp commented:
Great! Thanks for the changes, all looks good to me now. +1 from me.

TomDLT commented:
Thanks again for the review!

FYI I am working on a multinomial version of SAG, but it will be in another PR.

agramfort commented:
LGTM too

amueller commented on doc/whats_new.rst:
```diff
@@ -156,6 +158,12 @@ Enhancements
      features. By `Arnaud Joly`_.
 
+   - Added optional parameter ``random_state`` in :class:`linear_model.Ridge`
+     , to set the seed of the pseudo random generator used in ``sag`` solver.By `Tom Dupre la Tour`_.
```
space before "By" ;)

amueller commented:
@ogrisel we want this in the release, right?

amueller commented:
Great work everybody :)

ogrisel commented:
> @ogrisel we want this in the release, right?

I am not opposed to have it in :)

ogrisel commented:
> FYI I am working on a multinomial version of SAG, but it will be in another PR.

Would be great to consider adding support for `sample_weight` to `LogisticRegression` and `LogisticRegressionCV` as well while you are at it.

ogrisel commented:
This PR need a rebase on top of the current master.

amueller commented:
I'll rebase, squash and merge in a bit unless anyone complains.

amueller commented:
Pushed as 94eb61960a13f84a8be0dd527012f929172a0d60. Thanks for the great work!

agramfort commented:
:beers: !

ogrisel commented:
:beers:!

dsullivan7 commented:
Awesome!!

TomDLT commented:
Nice !

fabianp commented:
Yeah! @TomDLT deserves extra kudos for patience and perseverance :-)

TomDLT commented:
Thanks :)
